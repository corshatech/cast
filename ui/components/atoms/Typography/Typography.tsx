import clsx from 'clsx';
import React from 'react';

export type TypographyProps = {
  component?: 'h1' | 'h2' | 'h3' | 'h4' | 'p';
  variant?: 'h1' | 'h2' | 'h3' | 'h4' | 'body1' | 'body2' | 'body3' | 'body4';
  children: React.ReactNode[] | string;
};

const classes = {
  variant: (val: TypographyProps['variant']) => {
    switch (val) {
      case 'h1':
        return 'text-h1 font-poppins font-bold';
      case 'h2':
        return 'text-h2 font-poppins font-bold';
      case 'h3':
        return 'text-h3 font-poppins font-bold';
      case 'h4':
        return 'text-h4 font-poppins font-semibold';
      case 'body1':
        return 'text-body1 font-nunito';
      case 'body2':
        return 'text-body2 font-nunito';
      case 'body3':
        return 'text-body3 font-nunito';
      case 'body4':
        return 'text-body4 font-nunito';
      default:
        return 'text-body1 font-nunito';
    }
  },
};

const FontRender = ({
  component,
  children,
  variant,
}: TypographyProps): JSX.Element => {
  switch (component) {
    case 'h1':
      return (
        <h1 className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </h1>
      );
    case 'h2':
      return (
        <h2 className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </h2>
      );
    case 'h3':
      return (
        <h3 className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </h3>
      );
    case 'h4':
      return (
        <h4 className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </h4>
      );
    case 'p':
      return (
        <p className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </p>
      );
    default:
      return (
        <span className={clsx('leading-6', classes.variant(variant))}>
          {children}
        </span>
      );
  }
};

export const Typography = ({ variant, component, children }: TypographyProps) =>
  !!component ? (
    <FontRender variant={variant} component={component}>
      {children}
    </FontRender>
  ) : (
    <span className={clsx('leading-6', classes.variant(variant))}>
      {children}
    </span>
  );
