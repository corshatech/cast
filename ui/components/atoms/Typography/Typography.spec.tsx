import renderer from 'react-test-renderer';
import { Typography, TypographyProps } from './Typography';

const cases: [TypographyProps['variant'], TypographyProps['component']][] = [
  ['h1', 'h1'],
  ['h1', undefined],
  ['h2', 'h2'],
  ['h2', undefined],
  ['h3', 'h3'],
  ['h3', undefined],
  ['h4', 'h4'],
  ['h4', undefined],
  ['body1', 'p'],
  ['body1', undefined],
  ['body2', 'p'],
  ['body2', undefined],
  ['body3', 'p'],
  ['body3', undefined],
  ['body4', 'p'],
  ['body4', undefined],
];

describe('Typography', () => {
  test.each(cases)(
    'given %p as variant and %p as size, matches snapshot',
    (variant, component) => {
      const tree = renderer
        .create(
          <Typography variant={variant} component={component}>
            Typography Element
          </Typography>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    },
  );
});
