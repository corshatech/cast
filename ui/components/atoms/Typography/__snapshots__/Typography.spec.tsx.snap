// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography given "body1" as variant and "p" as size, matches snapshot 1`] = `
<p
  className="leading-6 text-body1 font-nunito"
>
  Typography Element
</p>
`;

exports[`Typography given "body1" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-body1 font-nunito"
>
  Typography Element
</span>
`;

exports[`Typography given "body2" as variant and "p" as size, matches snapshot 1`] = `
<p
  className="leading-6 text-body2 font-nunito"
>
  Typography Element
</p>
`;

exports[`Typography given "body2" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-body2 font-nunito"
>
  Typography Element
</span>
`;

exports[`Typography given "body3" as variant and "p" as size, matches snapshot 1`] = `
<p
  className="leading-6 text-body3 font-nunito"
>
  Typography Element
</p>
`;

exports[`Typography given "body3" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-body3 font-nunito"
>
  Typography Element
</span>
`;

exports[`Typography given "body4" as variant and "p" as size, matches snapshot 1`] = `
<p
  className="leading-6 text-body4 font-nunito"
>
  Typography Element
</p>
`;

exports[`Typography given "body4" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-body4 font-nunito"
>
  Typography Element
</span>
`;

exports[`Typography given "h1" as variant and "h1" as size, matches snapshot 1`] = `
<h1
  className="leading-6 text-h1 font-poppins font-bold"
>
  Typography Element
</h1>
`;

exports[`Typography given "h1" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-h1 font-poppins font-bold"
>
  Typography Element
</span>
`;

exports[`Typography given "h2" as variant and "h2" as size, matches snapshot 1`] = `
<h2
  className="leading-6 text-h2 font-poppins font-bold"
>
  Typography Element
</h2>
`;

exports[`Typography given "h2" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-h2 font-poppins font-bold"
>
  Typography Element
</span>
`;

exports[`Typography given "h3" as variant and "h3" as size, matches snapshot 1`] = `
<h3
  className="leading-6 text-h3 font-poppins font-bold"
>
  Typography Element
</h3>
`;

exports[`Typography given "h3" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-h3 font-poppins font-bold"
>
  Typography Element
</span>
`;

exports[`Typography given "h4" as variant and "h4" as size, matches snapshot 1`] = `
<h4
  className="leading-6 text-h4 font-poppins font-semibold"
>
  Typography Element
</h4>
`;

exports[`Typography given "h4" as variant and undefined as size, matches snapshot 1`] = `
<span
  className="leading-6 text-h4 font-poppins font-semibold"
>
  Typography Element
</span>
`;
